private/bufpkg/bufmodule/bufmoduleapi/uploader.go

```diff
@@ -107,6 +108,38 @@ func (a *uploader) Upload(
 	if err != nil {
 		return nil, err
 	}
+	// Only push named modules to the registry. Any dependencies for named modules must have a name.
+	// Local unnamed modules can be skipped if the UploadWithSkipUnnamed option is set.
+	contentModules, err = slicesext.FilterError(contentModules, func(module bufmodule.Module) (bool, error) {
+		moduleName := module.ModuleFullName()
+		if moduleName == nil {
+			if uploadOptions.SkipUnnamed() {
+				a.logger.Warn("Skipping unnamed module", zap.String("module", module.OpaqueID()))
+				return false, nil
+			}
+			return false, fmt.Errorf("A name must be specified in buf.yaml to push module %q.", module.OpaqueID())
+		}
+		deps, err := module.ModuleDeps()
+		if err != nil {
+			return false, err
+		}
+		if err := slicesext.Reduce(deps, func(err error, dep bufmodule.ModuleDep) error {
+			if moduleName := dep.ModuleFullName(); moduleName == nil {
+				return multierr.Append(err, fmt.Errorf("dependency %q does not have a name", dep.OpaqueID()))
+			}
+			return err
+		}, nil); err != nil {
+			return false, fmt.Errorf("All dependencies for module %s must have a name: %w.", moduleName.String(), err)
```

Comment:
Using multierr here is going to result in a not-fantastic actual error message for the end-user, it'd be better if we were able to print the string `all dependencies for module %q must be named but modules %s had no name", moduleName.String(), strings.Join(allDepModuleOpaqueIDs, ", "`

Reply:
This sounds right - thanks for being consistent!